src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp

```diff
@@ -345,86 +200,192 @@ uint G1ConcurrentRefine::max_num_threads() {
+                                                     double goal_ms) {
+  size_t minimum = minimum_pending_cards_target();
+  if ((processed_logged_cards < minimum) || (logged_cards_scan_time_ms == 0.0)) {
+    log_debug(gc, ergo, refine)("Unchanged pending cards target: %zu",
+                                _pending_cards_target);
+    return;
+  }
+
+  // Base the pending cards budget on the measured rate.
+  double rate = processed_logged_cards / logged_cards_scan_time_ms;
+  size_t budget = static_cast<size_t>(goal_ms * rate);
+  // Deduct predicted cards in thread buffers to get target.
+  size_t new_target = budget - MIN2(budget, predicted_thread_buffer_cards);
+  // Add some hysterisis with previous values.
+  if (is_pending_cards_target_initialized()) {
+    new_target = (new_target + _pending_cards_target) / 2;
+  }
+  // Apply minimum target.
+  new_target = MAX2(new_target, minimum_pending_cards_target());
+  _pending_cards_target = new_target;
+  log_debug(gc, ergo, refine)("New pending cards target: %zu", new_target);
+}
+
+void G1ConcurrentRefine::adjust_after_gc(double logged_cards_scan_time_ms,
+                                         size_t processed_logged_cards,
+                                         size_t predicted_thread_buffer_cards,
+                                         double goal_ms) {
+  if (!G1UseConcRefinement) return;
+  update_pending_cards_target(logged_cards_scan_time_ms,
+                              processed_logged_cards,
+                              predicted_thread_buffer_cards,
+                              goal_ms);
+  if (_thread_control.max_num_threads() == 0) {
+    // If no refinement threads then the mutator threshold is the target.
+    _dcqs.set_max_cards(_pending_cards_target);
+  } else {
+    // Provisionally make the mutator threshold unlimited, to be updated by
+    // the next periodic adjustment.  Because card state may have changed
+    // drastically, record that adjustment is needed and kick the primary
+    // thread, in case it is waiting.
+    _dcqs.set_max_cards(SIZE_MAX);
+    _needs_adjust = true;
+    if (is_pending_cards_target_initialized()) {
+      _thread_control.activate(0);
     }
-  } else if (logged_cards_scan_time < goal_ms &&
-             processed_logged_cards > green) {
-    green = static_cast<size_t>(MAX2(green * inc_k, green + 1.0));
-    green = MIN2(green, max_green_zone);
   }
-  return green;
 }
 
-static size_t calc_new_yellow_zone(size_t green, size_t min_yellow_size) {
-  size_t size = green * 2;
-  size = MAX2(size, min_yellow_size);
-  return MIN2(green + size, max_yellow_zone);
+// Wake up the primary thread less frequently when the time available until
+// the next GC is longer.  But don't increase the wait time too rapidly.
+// This reduces the number of primary thread wakeups that just immediately
+// go back to waiting, while still being responsive to behavior changes.
+static uint64_t compute_adjust_delay(double available_ms) {
```

Comment:
`delay` and `wait` used interchangeably, maybe better to choose one

Reply:
s/delay/wait_time_ms/